internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
+// or not the service exited with a success exit code (0).
+func getAction(config *plan.Service, success bool) (action plan.ServiceAction, onType string) {
+	switch {
+	case !success && config.OnFailure != "":
+		return config.OnFailure, "on-failure"
+	case success && config.OnSuccess != "":
+		return config.OnSuccess, "on-success"
+	default:
+		onExit := config.OnExit
+		if onExit == "" {
+			onExit = plan.ActionRestart // default for "on-exit"
+		}
+		return onExit, "on-exit"
 	}
+}
 
-	active, ok := m.services[req.Name]
-	if !ok {
-		// Already stopped
-		m.state.Lock()
-		task.Logf("Service %q already stopped.", req.Name)
-		m.state.Unlock()
-		return nil
+// sendSignal sends the given signal to a running service.
+func (s *service) sendSignal(signal string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateStarting, stateRunning:
+		err := syscall.Kill(-s.cmd.Process.Pid, unix.SignalNum(signal))
+		if err != nil {
+			return err
+		}
+
+	case stateBackoffWait, stateTerminating, stateKilling, stateStopped:
+		return fmt.Errorf("cannot send signal while service is stopped or stopping")
+
+	default:
+		return fmt.Errorf("sendSignal invalid in state %q", s.state)
 	}
-	cmd := active.cmd
+	return nil
+}
 
-	releasePlan()
+// stop is called to stop a running (and backing off) service.
+func (s *service) stop() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
 
-	syscall.Kill(-cmd.Process.Pid, syscall.SIGTERM)
-	//cmd.Process.Signal(syscall.SIGTERM)
+	switch s.state {
+	case stateRunning:
+		// First send SIGTERM to try to terminate it gracefully.
+		err := syscall.Kill(-s.cmd.Process.Pid, syscall.SIGTERM)
+		if err != nil {
+			logger.Noticef("cannot send SIGTERM to process: %v", err)
+		}
+		s.stopped = make(chan error, 1)
+		s.transition(stateTerminating)
+		time.AfterFunc(killWait, func() { logError(s.terminateTimeElapsed()) })
 
-	// TODO Make these timings configurable in the layer itself.
-	kill := time.After(killWait)
-	fail := time.After(failWait)
-	for {
-		time.Sleep(250 * time.Millisecond)
-		select {
-		case <-kill:
-			syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
-			//cmd.Process.Signal(syscall.SIGKILL)
-		case <-fail:
-			return fmt.Errorf("process still runs after SIGTERM and SIGKILL")
-		case <-active.done:
-			releasePlan, err := m.acquirePlan()
-			if err == nil {
-				if m.services[req.Name].cmd == cmd {
-					delete(m.services, req.Name)
-				}
-				releasePlan()
-			}
-			return nil
+	case stateBackoffWait:
+		s.transition(stateStopped)
+
+	default:
+		return fmt.Errorf("stop invalid in state %q", s.state)
+	}
+	return nil
+}
+
+// backoffWaitElapsed is called when the current backoff's timer has elapsed,
+// to restart the service.
+func (s *service) backoffWaitElapsed() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateBackoffWait:
+		logger.Debugf("Restarting service %q", s.config.Name)
```

Comment:
Shouldn't that be a nice and more general log message (non-debug)? Or do we have that elsewhere already?

Reply:
Yeah, you're right. I've made this a `logger.Noticef` that happens whenever the service starts. That's an important enough event!